homework/shared_ptr/shared_ptr-ut.cpp

```diff
@@ -0,0 +1,190 @@
+#include <gtest/gtest.h>
+
+#include <string>
+#include <vector>
+
+#include "shared_ptr.hpp"
+#include "weak_ptr.hpp"
+
+constexpr int testValue = 10;
+constexpr int secondTestValue = 666;
+
+struct SharedPointerTest : ::testing::Test {
+    SharedPointerTest()
+        : sharedPtr(new int(testValue)) {}
+    coders::shared_ptr<int> sharedPtr;
+};
+
+TEST_F(SharedPointerTest, ShouldCreateNullptrSharedPointerToIntegerUsingDefaultConstructor) {
+    // Given
+    size_t sharedPtrCnt = 1;
+
+    // When
+    coders::shared_ptr<int> emptySharedPtr{};
+
+    // Then
+    ASSERT_FALSE(emptySharedPtr);
+    ASSERT_EQ(emptySharedPtr.get(), nullptr);
+    ASSERT_EQ(emptySharedPtr.operator->(), nullptr);
+    ASSERT_EQ(emptySharedPtr.use_count(), sharedPtrCnt);
+}
+
+TEST_F(SharedPointerTest, ShouldCreateSharedPointerToIntegerUsingBasicConstructor) {
+    // Given
+    size_t sharedPtrCnt = 1;
+
+    // When
+    // Then
+    ASSERT_TRUE(sharedPtr);
+    ASSERT_EQ(sharedPtr.use_count(), sharedPtrCnt);
+    ASSERT_EQ(*sharedPtr.get(), testValue);
+}
+
+TEST_F(SharedPointerTest, ShouldCreateSharedPointerUsingCopyConstructor) {
+    // Given
+    size_t sharedPtrCnt = 1;
+    ASSERT_EQ(sharedPtr.use_count(), sharedPtrCnt);
+    ASSERT_EQ(*sharedPtr.get(), testValue);
+
+    // When
+    auto newShared(sharedPtr);
+    ++sharedPtrCnt;
+
+    // Then
+    ASSERT_TRUE(newShared);
+    ASSERT_EQ(newShared.use_count(), sharedPtrCnt);
+    ASSERT_EQ(*newShared.get(), testValue);
+}
+
+TEST_F(SharedPointerTest, ShouldCreateSharedPointerUsingCopyAssignmentOperator) {
+    // Given
+    size_t sharedPtrCnt = 1;
+    ASSERT_EQ(sharedPtr.use_count(), sharedPtrCnt);
+    ASSERT_EQ(*sharedPtr.get(), testValue);
+
+    // When
+    coders::shared_ptr<int> newShared;
+    newShared = sharedPtr;
+    ++sharedPtrCnt;
+
+    // Then
+    ASSERT_TRUE(newShared);
+    ASSERT_EQ(newShared.use_count(), sharedPtrCnt);
+    ASSERT_EQ(*newShared.get(), testValue);
+}
+
+TEST_F(SharedPointerTest, ShouldCreateSharedPointerUsingMoveConstructor) {
+    // Given
+    size_t sharedPtrCnt = 1;
+    ASSERT_EQ(sharedPtr.use_count(), sharedPtrCnt);
+    ASSERT_EQ(*sharedPtr.get(), testValue);
+
+    // When
+    coders::shared_ptr<int> newShared(std::move(sharedPtr));
+
+    // Then
+    ASSERT_FALSE(sharedPtr);
+    ASSERT_TRUE(newShared);
+    ASSERT_EQ(newShared.use_count(), sharedPtrCnt);
+    ASSERT_EQ(*newShared.get(), testValue);
+}
+
+TEST_F(SharedPointerTest, ShouldCreateSharedPointerUsingMoveAssignmentOperator) {
+    // Given
+    size_t sharedPtrCnt = 1;
+    ASSERT_EQ(sharedPtr.use_count(), sharedPtrCnt);
+    ASSERT_EQ(*sharedPtr.get(), testValue);
+
+    // When
+    coders::shared_ptr<int> newShared;
```

Comment:
please do not use the default constructor before copy/move into this object. Nothing can leak here.
```suggestion
    coders::shared_ptr<int> newShared{new int{0}};
```